README.md
[![Built on Travis](https://secure.travis-ci.org/mveitas/loggly-rsyslog.png?branch=master)](http://travis-ci.org/mveitas/loggly-rsyslog)

Loggly rsyslog Cookbook
================
Installs and configures rsyslog for use with [Loggly](http://loggly.com). This cookbook was built upon the work from an existing cookbook, https://github.com/kdaniels/loggly-rsyslog.

Requirements
------------
- Chef 11 or higher
- **Ruby 1.9.3 or higher**
- rsyslog is exists on the system

Platform
--------
Tested against Ubuntu 12.04

Data Bags
---------
This cookbook depends on the use of **encrypted data bags** to store the token to be used. For more information about data bags see the [Chef Data Bags](http://docs.opscode.com/essentials_data_bags.html) documentation. The data bag needs to be named `loggly` and contains an item `token`:

```
{
    "id": "token",
    "token": "<your token goes here>"
}
```

Attributes
----------
* `node['loggly']['tags']` - A list of event tags to apply to a message (https://www.loggly.com/docs/tags/) (optional)

* `node['loggly']['log_files']` - A list of files rsyslog should monitor. (optional). Below is an example
of a hash used to describe a file to monitor.

  ```
    {
        :filename => "/var/log/filename.log",
        :tag => "tag you want for this logfile",
        :statefile => "unique-name-for-statefile"
    }
    ```

* `node['loggly']['log_dirs']` - A list of directories to monitor (optional). The rsyslog.conf template will create an [imfile](http://www.rsyslog.com/doc/imfile.html) block for each file ending in '.log' in that directory. Each logdir in the list is of the format:
    ```
    {
        :directory => "/var/log/directory",
        :tag => "tag for all files in this directory"
    }
    ```

* `node['loggly']['tls']['enabled']` - Set to true if communication to the remote service should use TLS (defaults to true)
* `node['loggly']['tls']['cert_path']` - Directory where the loggly certificate should be placed
* `node['loggly']['tls']['cert_url']` - Url to the loggly.com certificate
* `node['loggly']['tls']['cert_checksum']` - Cchecksum of the loggly.com certificate
* `node['loggly']['tls']['intermediate_cert_url']` - Url to the intermediate certificate
* `node['loggly']['tls']['intermediate_cert_checksum']` - Checksum of the intermediate certificate

* `node['loggly']['rsyslog']['host']` - Name of the remote loggly syslog host
* `node['loggly']['rsyslog']['port']` - Port of the remote loggly syslog host (defaults to 514 and if TLS is enabled to 6514)

The following are attributes that manage some standard rsyslog configuration. See the rsyslog docs for more information on each of these.
* `node['loggly']['rsyslog']['repeat_msg']` - Set the RepeatedMsgReduction configuration value
* `node['loggly']['rsyslog']['file_owner']` - Set the FileOwnder configuration value
* `node['loggly']['rsyslog']['file_group']` - Set the FileGroup configuration value
* `node['loggly']['rsyslog']['file_create_mode']` - Set the FileCreateMode configuraton value
* `node['loggly']['rsyslog']['dir_create_mode']` - Set the DirCreateMode configuraton value
* `node['loggly']['rsyslog']['umask']` - Set the Umask configuration value
* `node['loggly']['rsyslog']['priv_drop_to_user']` - Set the PrivDropToUser configuration value
* `node['loggly']['rsyslog']['priv_drop_to_group']` - Set the PrivDropToGroup configuration value
* `node['loggly']['rsyslog']['work_directory']` - Set the WorkDirectory configuration value

Recipes
-------
Include the default recipe in the run list or a cookbook. The rsyslog service will restart after changes to the rsyslog.conf are made.

Running with Vagrant
--------------------
The create_data_bag.rb is a helper script that can be used to create the data bag for loggly for use with Vagrant and Chef Solo. The script expects a single arguement, the value of the loggly token.

Add the following two lines into your Vagrantfile in the chef_solo provisioner configuration:

```
chef.data_bags_path = './data_bags'
chef.encrypted_data_bag_secret_key_path = './encrypted_data_bag_secret'
```

License & Authors
-----------------
- Author: Matt Veitas <mveitas@gmail.com>

```text
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```



CHANGELOG.md
loggly-rsyslog Cookbook CHANGELOG
==========================
This file is used to list changes made in each version of the loggly-rsyslog cookbook.

v2.0.0
------------------
- Use an encrypted databag to retrieve the loggly token instead of a node attribute

v1.0.1
-------------------
- Set rsyslog configuration values to be configurable via attributes

v1.0.0 (1/25/2014)
-------------------
- Initial cookbook version
- Support for sending messages using TLS
- Configuration for monitoring a list of files
- Configuration for monitoring a list of directories


spec/tls_spec.rb
require 'spec_helper'

describe 'loggly-rsyslog::tls' do
  
  let(:chef_run) do
    ChefSpec::Runner.new do |node|
      node.set['loggly']['token'] = 'some_token_value'
    end.converge(described_recipe)
  end

  it 'uses 6514 as the port' do
    expect(chef_run.node['loggly']['rsyslog']['port']).to eq 6514
  end

  it 'does not install the rsyslog-gnutls package' do
    expect(chef_run).to install_package('rsyslog-gnutls')    
  end

  it 'creates a directory for the certificate' do
    expect(chef_run).to create_directory('/etc/rsyslog.d/keys/ca.d').with(
      owner: 'root',
      group: 'syslog',
      mode: 0750
    )
  end

  it 'installs the rsyslog-gnutls package' do
    expect(chef_run).to install_package('rsyslog-gnutls')
  end

  it 'downloads the certifcates' do
    expect(chef_run).to create_remote_file('download intermediate cert').with(
      path: "/var/chef/cache/sf_bundle.crt"
    )
    expect(chef_run).to create_remote_file('download loggly.com cert').with(
      path: "/var/chef/cache/loggly.com.crt"
    )
  end

  it 'creates the loggly certificate' do
    expect(chef_run).to run_bash('bundle certificate')
  end
  
  it 'loads the imfile module when log directories are not empty' do
    pending("Need to figure out stubbing of File.exists? as it used in many places inside Chef")
  end

end


spec/spec_helper.rb
require 'chefspec'

at_exit { ChefSpec::Coverage.report! }

spec/default_spec.rb
require 'spec_helper'

describe 'loggly-rsyslog::default' do

  before do
    Chef::EncryptedDataBagItem.stub(:load).with('loggly', 'token').and_return(
      { 'id' => 'token', 'token' => 'abc123' })
  end

  let(:chef_run) do
    ChefSpec::Runner.new do |node|
      node.set['loggly']['token'] = 'some_token_value'
    end.converge(described_recipe)
  end

  context 'when the loggly token is not set' do
    before do
      Chef::EncryptedDataBagItem.stub(:load).with('loggly', 'token').and_return(nil)
    end

    let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

    it 'raises an error' do
      expect {
        chef_run
      }.to raise_error
    end
  end

  context 'when rsyslog tls is disabled' do
    let(:chef_run) do
      ChefSpec::Runner.new do |node|
        node.set['loggly']['token'] = 'some_token_value'
      end.converge(described_recipe)
    end

    before do
      chef_run.node.set['loggly']['tls']['enabled'] = false
      chef_run.converge(described_recipe)
    end

    it 'does not include the tls recipe' do
      expect(chef_run).to_not include_recipe "loggly::tls"
    end

    it 'uses 514 as the port' do
      expect(chef_run.node['loggly']['rsyslog']['port']).to eq 514
    end

  end

  it 'contains the correct TLS configuration settings' do
    expect(chef_run).to render_file('/etc/rsyslog.conf').with_content(/^\$DefaultNetstreamDriverCAFile \/etc\/rsyslog.d\/keys\/ca.d\/loggly_full.crt/)    
    expect(chef_run).to render_file('/etc/rsyslog.conf').with_content(/^\$ActionSendStreamDriver gtls/)    
    expect(chef_run).to render_file('/etc/rsyslog.conf').with_content(/^\$ActionSendStreamDriverMode 1/)    
    expect(chef_run).to render_file('/etc/rsyslog.conf').with_content(/^\$ActionSendStreamDriverAuthMode x509\/name/)
    expect(chef_run).to render_file('/etc/rsyslog.conf').with_content(/^\$ActionSendStreamDriverPermittedPeer \*\.loggly.com/)    
  end

  it 'notifies the rsyslog service to restart' do
    rsyslog_template = chef_run.find_resource(:template, '/etc/rsyslog.conf')
    expect(rsyslog_template).to notify('service[rsyslog]').to(:restart)
  end

  it 'creates loggly rsyslog template with no tags' do
    expect(chef_run).to create_template('/etc/rsyslog.conf').with(
      owner: 'root',
      group: 'root',
      variables: ({
        :tags => '',
        :monitor_files => false,
        :token => 'abc123'
      })
    )  
  end

  it 'creates loggly rsyslog template with tags' do
    chef_run.node.set['loggly']['tags'] = ['test', 'foo', 'bar']
    chef_run.converge(described_recipe)
    expect(chef_run).to create_template('/etc/rsyslog.conf').with(
      owner: 'root',
      group: 'root',
      variables: ({
        :tags => 'tag=\"test\" tag=\"foo\" tag=\"bar\"',
        :monitor_files => false,
        :token => 'abc123'  
      })
    )  
  end

  it 'loads the imfile module when log_files is not empty' do
    runner = ChefSpec::Runner.new do |node|
      node.set['loggly']['token'] = 'some_token_value'
      node.set['loggly']['log_files'] = 
      [ { filename: '/var/log/somefile', tag: 'sometag', statefile: 'somefile.state' },
        { filename: '/var/log/anotherfile', tag: 'anothertag', statefile: 'anotherfile.state' }
      ]
    end.converge(described_recipe)

    expect(runner).to create_template('/etc/rsyslog.conf').with(
      owner: 'root',
      group: 'root',
      variables: ({
        :tags => '',
        :monitor_files => true,
        :token => 'abc123'  
      })
    )

    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^.*\[abc123@41058 \] %msg%/)    

    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^\$ModLoad imfile/)

    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^\$InputFileName \/var\/log\/somefile/)    
    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^\$InputFileTag sometag\:/)
    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^\$InputFileStateFile somefile.state/)
  end

  it 'loads the imfile module when log_files is not empty' do
    runner = ChefSpec::Runner.new do |node|
      node.set['loggly']['token'] = 'some_token_value'
      node.set['loggly']['log_files'] = 
      [ { filename: '/var/log/somefile', tag: 'sometag', statefile: 'somefile.state' } ]
    end.converge(described_recipe)  

    expect(runner).to create_template('/etc/rsyslog.conf').with(
      variables: ({
        :tags => '',
        :monitor_files => true,
        :token => 'abc123'  
      })
    )

    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^\$ModLoad imfile/)
    expect(runner).to render_file('/etc/rsyslog.conf').with_content(/^\$InputFilePollInterval 10/)
  end

  {
    'RepeatedMsgReduction' => 'on',
    'FileOwner' => 'syslog',
    'FileGroup' =>'adm',
    'FileCreateMode' => '0640',
    'DirCreateMode' =>'0755',
    'Umask' => '0022',
    'PrivDropToUser' => 'syslog',
    'PrivDropToGroup' => 'syslog',
    'WorkDirectory' => '/var/spool/rsyslog'
  }.each do |k, v| 
    it 'sets the rsyslog #{k} configuration value' do
      expect(chef_run).to render_file('/etc/rsyslog.conf').with_content(/^\$#{k} #{v}/)
    end
  end

end


recipes/default.rb
#
# Cookbook Name:: loggly
# Recipe:: default
#
# Copyright (C) 2014 AgaMatrix
# 
# All rights reserved - Do Not Redistribute
#

loggly_token = Chef::EncryptedDataBagItem.load('loggly', 'token')['token']
raise "No token was found in the loggly databag." if loggly_token.nil?

service "rsyslog"

include_recipe "loggly-rsyslog::tls" if node['loggly']['tls']['enabled']

template '/etc/rsyslog.conf' do
  source 'loggly.conf.erb'
  owner 'root'
  group 'root'
  mode 0644
  variables({
    :monitor_files => !node['loggly']['log_files'].empty? || !node['loggly']['log_dirs'].empty?,
    :tags => node['loggly']['tags'].nil? || node['loggly']['tags'].empty? ? '' : "tag=\\\"#{node['loggly']['tags'].join("\\\" tag=\\\"")}\\\"",
    :token => loggly_token
  })
  notifies :restart, "service[rsyslog]", :immediate
end


recipes/tls.rb
#
# Cookbook Name:: loggly
# Recipe:: tls
#
# Copyright (C) 2014 AgaMatrix
# 
# All rights reserved - Do Not Redistribute
#

package 'rsyslog-gnutls' do
  action :install
end

cert_path = node['loggly']['tls']['cert_path']

directory cert_path do
  owner 'root'
  group 'syslog'
  mode 0750
  action :create
  recursive true
end

loggly_crt_path = "#{Chef::Config['file_cache_path']}/loggly.com.crt"
sf_bundle_path = "#{Chef::Config['file_cache_path']}/sf_bundle.crt"

remote_file 'download loggly.com cert' do
  owner 'root'
  group 'root'
  mode 0644
  path loggly_crt_path
  source node['loggly']['tls']['cert_url']
  checksum node['loggly']['tls']['cert_checksum']
end

remote_file 'download intermediate cert' do
  owner 'root'
  group 'root'
  mode 0644
  path sf_bundle_path
  source node['loggly']['tls']['intermediate_cert_url']
  checksum node['loggly']['tls']['intermediate_cert_checksum']
end
  
bash 'bundle certificate' do
  user 'root'
  cwd cert_path
  code <<-EOH
    cat {#{sf_bundle_path},#{loggly_crt_path}} > loggly_full.crt
  EOH
  not_if { ::File.exists?("#{node['loggly']['tls']['cert_path']}/loggly_full.crt") }
end
